### PR TITLE
ONNX export for index_select

### PR DESCRIPTION
**Pull request by liqunfu**
ONNX export for index_select

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -931,6 +931,12 @@ def selu(g, input):
 
 @parse_args('v', 'i', 'v')
 def index_select(g, self, dim, index):
+    if index.type().kind() == "CompleteTensorType" or index.type().kind() == "DimensionedTensorType":
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset9.py]: We could avoid relying on the tensor type here and use _maybe_get_const() instead.

index_const = sym_help._maybe_get_const(index, 'i')
    if not sym_help._is_value(index_const):
        index = g.op("Reshape", index, g.op("Constant", value_t=torch.LongTensor([-1])))

**Comment by lara-hdr** [comment]
maybe add tests in torch/onnx/test_onnx_opset.py too, to check the schema

File: test/onnx/test_onnx_opset.py [hunk 2]
```diff
@@ -206,6 +207,21 @@ def forward(self, x):
...
+
+        x = torch.randn(3, 4)
+
+        ops = [
+            {"op_name" : "Constant"},
+            {"op_name" : "Constant"},
+            {"op_name" : "Reshape"},
+            {"op_name" : "Gather", "attributes" : [{"name" : "axis", "i" : 1, "type" : 2}]}]
+        ops = {9 : ops}
+        check_onnx_opsets_operator(MyModule(), x, ops, opset_versions=[9])
```
Comment (houseroad) [inline test/onnx/test_onnx_opset.py]: this tests fail in CI. IMO, opset test is for checking some special structure of the graph. Do we have any particular reason to check the structure of the graph?  I think end to end tests should be good enough in this case.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -6,13 +6,71 @@
...
         self.assertTrue('onnxruntime' in sys.modules)
 
+    # There is a difference between PyTorch and ONNX/numpy w.r.t index_select by a scaler index.
+    # With an input tensor of rank r and a scaler index, PyTorch index_select returns a tensor of rank = r.
+    # However, corresponding op in ONNX(Gather) and numpy returns a tensor of rank = r - 1.
+    # To maintain a parity between ONNX and PyTorch, scaler index is converted to a 1D tensor
+    # before applying an ONNX gather op during ONNX export. 
+    # Following test_index_select_* tests are to confirm that equivalence between PyTorch and ONNX 
+    # is maintained in the above case.
+    def test_index_select_constant_scaler_index(self):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Feel free to delete the first test. (It was added to verify I have successfully installed the onnxruntime.) Now we have real test cases.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -6,13 +6,71 @@
...
+        class IndexSelectScalerIndexModel(torch.nn.Module):
+            def forward(self, x):
+                return torch.index_select(x, 1, torch.tensor(index))
+
+        model = IndexSelectScalerIndexModel()
+        x = torch.randn(3, 4)
+        pt_out = model(x)
+
+        f = io.BytesIO()
+        torch.onnx._export(model, (x,), f)
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Feel free to add some helper function (e.g. assertONNXModel(input, pytorch_model, expect_output)) like what we did in test_pytorch_onnx_caffe2.py. I think it will help you simplify the test file.

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -931,6 +931,15 @@ def selu(g, input):
 
 @parse_args('v', 'i', 'v')
 def index_select(g, self, dim, index):
+    index_const = sym_help._maybe_get_scalar(index)
+    if not sym_help._is_value(index_const):
+        index = g.op("Constant", value_t=torch.LongTensor([index_const]))
+    elif index.type().kind() == "CompleteTensorType" or index.type().kind() == "DimensionedTensorType":
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: This pattern is pretty common, can we extract it as a helper function?

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -931,6 +931,15 @@ def selu(g, input):
...
+        if index.type().dim() == 0:
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: dim() != 1?

**Review by houseroad — CHANGES_REQUESTED** [review]
Please make sure CIs pass, and address the comments.

File: test/onnx/test_onnx_opset.py (hunk 2 above)
Comment (liqunfu) [reply test/onnx/test_onnx_opset.py]: Good point. I removed this test.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (liqunfu) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: @lara-hdr had removed this test in a different PR.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 4 above)
Comment (liqunfu) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: ort tests now use a test and verify helper

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (liqunfu) [reply torch/onnx/symbolic_opset9.py]: we want to work with scalar case which is dim() == 0.

File: torch/onnx/symbolic_opset9.py (hunk 5 above)
Comment (liqunfu) [reply torch/onnx/symbolic_opset9.py]: helper function has been added and used in a few places.

**Review by houseroad — APPROVED** [review]
LGTM
